src/btree/bt_ret.c

```diff
@@ -142,8 +142,20 @@ __value_return(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE *upd)
 int
 __wt_key_return(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt)
 {
-	WT_RET(__key_return(session, cbt));
+	WT_CURSOR *cursor;
+
+	cursor = &cbt->iface;
 
+	/*
+	 * We may already have an internal key, in which case the cursor may
+	 * not be set up to get another copy (for example, when we rely on a
+	 * search-function result).
+	 */
+	F_CLR(cursor, WT_CURSTD_KEY_EXT);
+	if (!F_ISSET(cursor, WT_CURSTD_KEY_INT)) {
+		WT_RET(__key_return(session, cbt));
+		F_SET(cursor, WT_CURSTD_KEY_INT);
+	}
```

Comment:
This seems obscure to me -- we'll silently fail to return the current key if one is already set in the cursor.  Can we assert `!F_ISSET(cursor, WT_CURSTD_KEY_INT)` here instead?